src/puppetlabs/puppetdb/query_eng/engine.clj

```diff
@@ -33,432 +39,623 @@
+                             "producer_timestamp" {:type :timestamp
+                                                   :queryable? true
+                                                   :field :factsets.producer_timestamp}
+                             "environment" {:type :string
+                                            :queryable? true
+                                            :field :environments.name}}
+
+               :selection {:from [:factsets]
+                           :left-join [:environments
+                                       [:= :factsets.environment_id :environments.id]]}
+
                :alias "factsets"
-               :queryable-fields ["certname" "environment" "timestamp"
-                                  "producer_timestamp" "hash"]
                :entity :factsets
                :source-table "factsets"
-               :subquery? false
-               :supports-extract? false
-               :source "SELECT fact_paths.path, timestamp,
-                               COALESCE(fact_values.value_string,
-                                        fact_values.value_json,
-                                        CAST(fact_values.value_boolean as text)) as value,
-                               fact_values.value_integer as value_integer,
-                               fact_values.value_float as value_float,
-                               factsets.certname,
-                               factsets.hash,
-                               factsets.producer_timestamp,
-                               environments.name as environment,
-                               value_types.type
-                        FROM factsets
-                             INNER JOIN facts on factsets.id = facts.factset_id
-                             INNER JOIN fact_values on facts.fact_value_id = fact_values.id
-                             INNER JOIN fact_paths on facts.fact_path_id = fact_paths.id
-                             INNER JOIN value_types on fact_values.value_type_id = value_types.id
-                             LEFT OUTER JOIN environments on factsets.environment_id = environments.id
-                        WHERE depth = 0
-                        ORDER BY factsets.certname"}))
+               :subquery? false}))
 
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 ;;; Conversion from plan to SQL
 
+(defn queryable-fields
+  "Returns a list of queryable fields from a query record.
+
+  These are fields with the setting :queryable? set to true."
+  [{:keys [projections]}]
+  (->> projections
+       (remove (comp (complement :queryable?) val))
+       keys
+       (into [])
+       sort))
+
+(defn projectable-fields
+  "Returns a list of projectable fields from a query record.
+
+   Fields marked as :query-only? true are unable to be projected and thus are
+   excluded."
+  [{:keys [projections]}]
+  (->> projections
+       (remove (comp :query-only? val))
+       keys
+       (into [])
```

Comment:
I don't think you need this here. Sort doesn't need a vec and doesn't return a vec even if you give it one.